# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

**Prediction of Model Performance**

I predict that the **Random Forests Classifier** will perform better than the Logistic Regression model on this spam data.

**Justification:**

1. **Handling Complex Relationships:** The spam data likely contains complex relationships between features that are not linearly separable. Random Forest, being an ensemble method of multiple decision trees, is typically better at capturing these non-linear interactions compared to Logistic Regression, which assumes a linear relationship between the features and the target variable.

2. **Feature Interactions:** Spam detection often involves interactions between various features (like word frequencies, email length, etc.) that Random Forest can effectively capture. In contrast, Logistic Regression may require explicit feature engineering to capture such interactions.

3. **Overfitting Resistance:** Random Forests have a natural resistance to overfitting, especially if the dataset is large and diverse, due to the averaging of decisions from multiple trees. On the other hand, Logistic Regression might overfit if not regularized properly.

4. **Versatility with Various Data Types:** Random Forests are generally versatile with different types of data and distributions, which might be beneficial if the spam dataset contains a mix of numerical and categorical variables.

However, it's important to note that Logistic Regression could perform better if the relationship between the features and the target is indeed linear, or if the dataset size is not large enough for Random Forest to exhibit its full potential.

## Split the Data into Training and Testing Sets

In [7]:
# Create the labels set `y` and features DataFrame `X`

# Assuming the last column is the target variable
y = data['spam']  # Extracts the 'spam' column as the labels set
X = data.drop('spam', axis=1)  # Drops the 'spam' column to keep only the features


In [8]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_balance = y.value_counts()
print(label_balance)

0    2788
1    1813
Name: spam, dtype: int64


In [9]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


      word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
3683            0.00                0.0           0.00           0.0   
4412            0.71                0.0           0.71           0.0   
2584            0.00                0.0           0.91           0.0   
69              0.00                0.0           0.00           0.0   
1844            0.00                0.0           0.54           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
3683            0.0             0.0               0.0                0.00   
4412            0.0             0.0               0.0                0.00   
2584            0.0             0.0               0.0                0.45   
69              0.0             0.0               0.0                0.00   
1844            0.0             0.0               0.0                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
3683              0.0  

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [11]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [12]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [13]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [16]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(random_state=1)
logistic_model.fit(X_train_scaled, y_train)
train_score = logistic_model.score(X_train_scaled, y_train)
print(f"Training Score: {train_score}")

# Prepare to scale X_test (this step was not mentioned but is necessary)
X_test_scaled = scaler.transform(X_test)

Training Score: 0.9274456521739131


In [19]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_model.predict(X_test_scaled)

# Review the predictions
print("Some testing predictions:", testing_predictions[:5])


Some testing predictions: [0 0 0 1 0]


In [20]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, testing_predictions)
print(f"Accuracy Score: {accuracy}")

Accuracy Score: 0.9196525515743756


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [22]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

random_forest_model = RandomForestClassifier(random_state=1)
random_forest_model.fit(X_train_scaled, y_train)
rf_train_score = random_forest_model.score(X_train_scaled, y_train)
print(f"Random Forest Training Score: {rf_train_score}")


Random Forest Training Score: 0.9994565217391305


In [23]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_testing_predictions = random_forest_model.predict(X_test_scaled)

# Review the predictions
print("Some Random Forest testing predictions:", rf_testing_predictions[:5])


Some Random Forest testing predictions: [0 0 0 1 0]


In [24]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_accuracy = accuracy_score(y_test, rf_testing_predictions)
print(f"Random Forest Accuracy Score: {rf_accuracy}")

Random Forest Accuracy Score: 0.9565689467969598


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

### Model Performance Recap

- **Logistic Regression Model:**
  - Training Score: 0.9274
  - Testing Accuracy: 0.9197

- **Random Forest Classifier Model:**
  - Training Score: 0.9995
  - Testing Accuracy: 0.9566

### Comparison with Prediction

- **Initial Prediction:** My initial prediction was that the Random Forest Classifier would perform better due to its ability to capture complex and non-linear relationships within the data, along with its natural resistance to overfitting.

- **Actual Outcome:** The Random Forest Classifier indeed performed better, as indicated by a higher testing accuracy of 0.9566 compared to 0.9197 for the Logistic Regression model. This aligns with the expectation that ensemble methods like Random Forest can more effectively manage the intricacies within the data.

### Reflection

The results align with my initial prediction, showcasing the strength of ensemble methods, particularly in complex datasets where the interactions between features are not easily modeled by simpler algorithms like Logistic Regression. While the Logistic Regression model also performed well, the Random Forest Classifier demonstrated superior accuracy, likely due to its ability to construct a more nuanced understanding of the data through multiple decision trees.

However, it's noteworthy that the Random Forest model showed a much higher training score, suggesting a potential overfit to the training data, although this did not seem to adversely impact its testing accuracy in this case.

In future model selections, I will continue to weigh the benefits of more complex models like Random Forest against simpler ones like Logistic Regression, particularly in terms of their training behavior and how they generalize to unseen data. This experience reinforces the importance of testing multiple models to find the most effective approach for a given dataset.